gctree/phylip_parse.py

```diff
@@ -127,28 +127,72 @@ def parse_outfile(outfile, abundance_file=None, root="root"):
     return trees
 
 
-def disambiguate(tree):
-    """make random choices for ambiguous bases, respecting tree inheritance."""
-    # random seed based on newick string of tree
-    random.seed(tree.write(format=1))
-    ambiguous_dna_values["?"] = "GATC-"
-    sequence_length = len(tree.sequence)
+def disambiguate(tree: Tree, random_state=None) -> Tree:
+    """Randomly resolve ambiguous bases using a two-pass Sankoff Algorithm on
+    subtrees of consecutive ambiguity codes."""
+    bases = "AGCT-"
+    ambiguous_dna_values.update({"?": "GATC-", "-": "-"})
+    code_vectors = {
+        code: [
+            0 if base in ambiguous_dna_values[code] else float("inf") for base in bases
+        ]
+        for code in ambiguous_dna_values
+    }
+    cost_adjust = {
+        base: [int(not i == j) for j in range(5)] for i, base in enumerate(bases)
+    }
```

Comment:
I suggest moving all this to module scope to improve the efficiency of multiple calls to `disambiguate`.